src/PluginManagerInterface.php

```diff
@@ -13,17 +15,53 @@
  * A plugin manager is a specialized service locator used to create homogeneous objects
  *
  * @template InstanceType
+ * @psalm-import-type ServiceManagerConfigurationType from ConfigInterface
  */
 interface PluginManagerInterface extends ServiceLocatorInterface
 {
     /**
      * Validate an instance
      *
-     * @return void
      * @throws InvalidServiceException If created instance does not respect the
      *     constraint on type imposed by the plugin manager.
      * @throws ContainerExceptionInterface If any other error occurs.
      * @psalm-assert InstanceType $instance
      */
-    public function validate(mixed $instance);
+    public function validate(mixed $instance): void;
+
+    public function has(string $id): bool;
+
+    /**
+     * @param class-string<InstanceType>|string $id Service name of plugin to retrieve.
+     * @psalm-return ($id is class-string<InstanceType> ? InstanceType : mixed)
+     * @throws Exception\ServiceNotFoundException If the manager does not have
+     *     a service definition for the instance, and the service is not
+     *     auto-invokable.
+     * @throws InvalidServiceException If the plugin created is invalid for the
+     *     plugin context.
+     */
+    public function get(string $id): mixed;
+
+    /**
+     * Build a service by its name, using optional options (such services are NEVER cached).
+     *
+     * @param  string|class-string<InstanceType> $name
+     * @psalm-return ($name is class-string<InstanceType> ? InstanceType : mixed)
+     * @throws Exception\ServiceNotFoundException If no factory/abstract
+     *     factory could be found to create the instance.
+     * @throws Exception\ServiceNotCreatedException If factory/delegator fails
+     *     to create the instance.
+     * @throws ContainerExceptionInterface If any other error occurs.
+     */
+    public function build(string $name, ?array $options = null): mixed;
+
+    /**
+     * @param ServiceManagerConfigurationType $config
+     * @throws ContainerModificationsNotAllowedException If the allow override flag has been toggled off, and a
+     *                                                   service instanceexists for a given service.
+     * @throws InvalidServiceException If an instance passed in the `services` configuration is invalid for the
+     *                                 plugin manager.
+     * @throws CyclicAliasException If the configuration contains aliases targeting themselves.
+     */
+    public function configure(array $config): static;
```

Comment:
IMO `configure()` shouldn't be part of the public API of a `PluginManagerInterface`, since the plugin manager is likely going to become a thin wrapper around a more complex and potentially completely immutable container.

Was it introduced on purpose here? What are the implication if we drop this?

Reply:
The various cache adapters are adding themselves to the CacheAdapterPluginManager via delegator factories. Since every plugin manager was extending servicemanager, that method was available before even tho it was nowhere declared. But maybe we can drop it from the Interface as usually the concrete pluginmanager is being fetched anyways.

Reply:
Yeah, I'd prefer typing `ServiceManager` there, instead of typing against the interface then :+1:

Reply:
But the delegators are not able to Access the SM directly. Due to the removal of the inheritance, there is no way to modify the SM from the delegators - thats why these methods were kept.

Reply:
Even then, I'd say that this is a design smell: you cannot assume that the container can be modified post-instantiation, while this method addition makes that a requirement of any implementation.

Reply:
Yah, I've removed the method from the interface while keeping it within the `AbstractPluginManager`. 

Are we fine with creating a frozen `ServiceManager` concept for the next major version rather than adding it to v4? 
v4 is overdue for years now and it feels like something no1 prepared in the last years and thus, at least for now, it should be fine with going the current direction. This will also avoid having too many refactoring requirements to consumers for this version.

Reply:
I do not think immutability should be dropped in the mix here. Make the move to PSR container as painless as it can be first.